### PR TITLE
Refine the API of BlockDevice trait

### PR DESCRIPTION
**Pull request by liqinggd**

### CODE REVIEW

File: services/comps/block/src/request_queue.rs [hunk 1]
```diff
@@ -7,24 +7,113 @@ use super::{
     id::Sid,
 };
 
-/// Represents the software staging queue for the `BioRequest` objects.
-pub trait BioRequestQueue {
+use aster_frame::sync::{Mutex, WaitQueue};
+
+/// A simple block I/O request queue with a single queue.
```
Comment (tatetian) [inline services/comps/block/src/request_queue.rs:12]: This sentence can be boiled down to "a queue with a queue". This expression is illogical. Consider rewriting this way.

```
/// A simple block I/O request queue backed by one internal FIFO queue.
```

In addition, the Rust doc should mention that the queue support merging the new request with the front request.

File: services/comps/block/src/request_queue.rs [hunk 2]
```diff
@@ -7,24 +7,113 @@ use super::{
...
+            if request.can_merge(&bio) {
+                request.merge_bio(bio);
+                return Ok(());
+            }
+        }
+
+        let new_request = BioRequest::from(bio);
+        queue.push_front(new_request);
+        drop(queue);
+        self.inc_num_requests();
```
Comment (tatetian) [inline services/comps/block/src/request_queue.rs:57]: In certain race conditions between concurrent `enqueue` and `dequeue`, the counter may be decreased first, making the number underflow temporarily. This should be avoided.

File: services/comps/block/src/request_queue.rs (hunk 1 above)
Comment (liqinggd) [reply services/comps/block/src/request_queue.rs]: Done

File: services/comps/block/src/request_queue.rs (hunk 2 above)
Comment (liqinggd) [reply services/comps/block/src/request_queue.rs:59]: Yes the modification of `num_requests` should be controlled by the lock of  queue.

**Review by tatetian — APPROVED** [review]
LGTM. Thanks for the contribution!

**Pull request merged**